### PR TITLE
Add Inplace strategy (Output reuse Input Varbase) in dygraph

### PR DESCRIPTION
**Pull request by pangyoki**
### PR types

New features

### PR changes

APIs

### Describe

动态图支持Inplace策略：不新建输出VarBase，直接将输入VarBase传递给输出。与静态图已添加Inplace策略的op对齐。为与该策略相关的OP新增python API接口。

- 支持的OP
与静态图已支持Inplace策略的OP对齐，一共有38个。在这个PR中，为`squeeze2`、`unsqueeze2`、`reshape2`、`scatter`
`elu`、`relu`、`softmax`、`tanh` 8个OP添加了Inplace策略。其它OP后续会添加Inplace支持。

- API接口
为需要使用Inplace策略的OP新增python API接口，API名称为现有API名称加上下划线_，用户可选择是否使用这类API组网。
对上述8个OP，对应添加python API接口：`squeeze_`、`unsqueeze_`、`reshape_`、`scatter_`
`elu_`、`relu_`、`softmax_`、`tanh_` 。

- 生成的op function
在op_function_generator层面添加动态图Inplace机制，以squeeze2 op为例，添加Inplace策略后，生成的op function为`imperative_squeeze2_`。
主要改变为：
  - 检查需要做inplace的var是否为`stop_gradient=False`的叶子节点，如果是，则会报错。
  - inplace var的 `inplace_version` 增加1。
  - 输入Var直接传递给输出。
  ```
  std::tuple<std::shared_ptr<imperative::VarBase>,std::shared_ptr<imperative::VarBase>> imperative_squeeze2_(const py::handle& X_, const py::args& args)
  {
    
    auto X = CastPyHandleToVarBase("squeeze2", "X", 0, X_, false);
    framework::AttributeMap attrs;
    ConstructAttrMapFromPyArgs("squeeze2", 1, &attrs, args);
    {
      py::gil_scoped_release release;
      auto tracer = imperative::GetCurrentTracer();
      
      PADDLE_ENFORCE_EQ(
        X->IsLeaf() && !X->OverridedStopGradient(), false,
        platform::errors::InvalidArgument("Leaf Var (%s) that doesn't stop gradient can't use inplace strategy.", X->Name()));
      X->BumpInplaceVersion();
      VLOG(3) << "Var(" << X->Name() << ") uses Inplace Strategy.";
  
      imperative::NameVarBaseMap outs = {{"Out", {X}},{"XShape", {std::shared_ptr<imperative::VarBase>(new imperative::VarBase(tracer->GenerateUniqueName()))}}};
      imperative::NameVarBaseMap ins = {{"X", {X}}};
      
      tracer->TraceOp("squeeze2", ins, outs, attrs, {{"X", "Out"}});
      return std::make_tuple(outs["Out"][0],outs["XShape"][0]); 
    }   
  }
  ```

- `squeeze_()` API使用示例
  ```
  import paddle
  
  x = paddle.rand([5, 1, 10])
  output = paddle.squeeze_(x, axis=1)
  
  print(x.shape)  # [5, 10]
  print(output.shape)  # [5, 10]
  print(id(x) == id(output))  # True
  ```

- Inplace报错
    - `stop_gradient=False`的叶子节点
    ```
    import paddle
    
    a = paddle.rand([2,3,1])
    a.stop_gradient = False
    a.squeeze_()
    # ValueError: (InvalidArgument) Leaf Var (dygraph_tmp_2) that doesn't stop gradient can't use inplace strategy.
    ```

     - 反向执行计算错误，Inplace反向检测，实现见 PR #27901 。
    ```
    import paddle
    
    a = paddle.rand([2,3,1])
    a.stop_gradient = False
    
    b = a ** 2
    c = b ** 2
    b.squeeze_()
    
    c.sum().backward()
    # RuntimeError: (PermissionDenied) Tensor 'dygraph_tmp_2' used in gradient computation in grad op 'elementwise_pow_grad' has been modified by an inplace operation. Its version is 1 but the expected version is 0. Please fix your code to void calling an inplace operator after using the Tensor which will used in gradient computation
    ```

- 实现要点

  - op function generator中加入动态图使用Inplace策略的逻辑
  - 反向op执行时不会用Inplace策略，对inplace反向op，需要生成临时var
  - inplace反向构造网络时会出现递归问题：修改GradOpMaker中SetInput和SetOutput逻辑
  - inplace var会出现梯度聚合问题：将basic_engine中accumulator_的key加入grad_node

### CODE REVIEW

File: python/paddle/tensor/manipulation.py [hunk 1]
```diff
@@ -65,6 +70,12 @@
 ]
 
 
+def dygraph_inplace_in_static_mode(api_name):
```
Comment (zhiqiu) [inline python/paddle/tensor/manipulation.py]: Better name it "_print_warning_in_static_mode"

File: python/paddle/tensor/manipulation.py [hunk 2]
```diff
@@ -567,6 +578,40 @@ def squeeze(x, axis=None, name=None):
     return layers.squeeze(x, axis, name)
 
 
+def squeeze_(x, axis=None, name=None):
+    """
+    Inplace version of ``squeeze`` API. Please refer to doc of ``squeeze``.
+    Inplace strategy: Do not create a new output Tensor, output and input is the same Tensor.
```
Comment (zhiqiu) [inline python/paddle/tensor/manipulation.py]: As discussed, remove this line and write an introduction doc of Inplace.
Same for other APIs.

File: python/paddle/tensor/manipulation.py [hunk 3]
```diff
@@ -567,6 +578,40 @@ def squeeze(x, axis=None, name=None):
     return layers.squeeze(x, axis, name)
 
 
+def squeeze_(x, axis=None, name=None):
+    """
+    Inplace version of ``squeeze`` API. Please refer to doc of ``squeeze``.
```
Comment (zhiqiu) [inline python/paddle/tensor/manipulation.py]: Inplace version of ``squeeze`` API, the output Tensor will be inplaced with input ``x``.

File: python/paddle/tensor/manipulation.py [hunk 4]
```diff
@@ -567,6 +578,40 @@ def squeeze(x, axis=None, name=None):
...
+        .. code-block:: python
+
+            import paddle
+
+            x = paddle.rand([5, 1, 10])
+            output = paddle.squeeze_(x, axis=1)
+
+            print(x.shape)  # [5, 10]
+            print(output.shape)  # [5, 10]
+            print(id(x) == id(output))  # True
```
Comment (zhiqiu) [inline python/paddle/tensor/manipulation.py]: Better discuss with TPM about whether the example code is needed.

File: python/paddle/tensor/manipulation.py [hunk 5]
```diff
@@ -65,6 +70,12 @@
 ]
 
 
+def dygraph_inplace_in_static_mode(api_name):
+    warnings.warn(
+        "{}_() API only uses Inplace strategy in dygraph mode, the function is the same as {}() in static mode.".
+        format(api_name, api_name))
```
Comment (zhiqiu) [inline python/paddle/tensor/manipulation.py]: In static mode, {}_() is same as {}() and does not perform inplace operation.

File: paddle/fluid/imperative/tracer.cc [hunk 6]
```diff
@@ -130,7 +131,8 @@ paddle::framework::GarbageCollector* Tracer::MutableGarbageCollectorIfNotExists(
 
 void Tracer::TraceOp(const std::string& type, const NameVarBaseMap& ins,
                      const NameVarBaseMap& outs, framework::AttributeMap attrs,
-                     const platform::Place& place, bool trace_backward) {
+                     const platform::Place& place, bool trace_backward,
+                     const std::map<std::string, std::string>& inplace_map) {
```
Comment (zhiqiu) [inline paddle/fluid/imperative/tracer.cc:135]: Set default value for the new parameter, so the existing code needs less modification.
 `const std::map<std::string, std::string>& inplace_map = {}`

File: python/paddle/fluid/tests/unittests/test_squeeze_op.py [hunk 7]
```diff
@@ -98,13 +98,19 @@ def test_errors(self):
...
+        self.squeeze = paddle.squeeze
+
     def test_out(self):
         paddle.enable_static()
         with paddle.static.program_guard(paddle.static.Program(),
                                          paddle.static.Program()):
             data1 = paddle.static.data(
                 'data1', shape=[-1, 1, 10], dtype='float64')
-            result_squeeze = paddle.squeeze(data1, axis=[1])
+            result_squeeze = self.squeeze(data1, axis=[1])
```
Comment (zhiqiu) [inline python/paddle/fluid/tests/unittests/test_squeeze_op.py:113]: Why change this?

Comment (pangyoki) [reply python/paddle/fluid/tests/unittests/test_squeeze_op.py:113]: In order to allow `paddle.squeeze_()` to reuse this code. The following `API_TestStaticSqueeze_` class inherits this class.

File: paddle/fluid/imperative/tracer.cc (hunk 6 above)
Comment (pangyoki) [reply paddle/fluid/imperative/tracer.cc:135]: done

File: python/paddle/tensor/manipulation.py (hunk 5 above)
Comment (pangyoki) [reply python/paddle/tensor/manipulation.py]: done

File: python/paddle/tensor/manipulation.py (hunk 3 above)
Comment (pangyoki) [reply python/paddle/tensor/manipulation.py]: done

File: python/paddle/tensor/manipulation.py (hunk 2 above)
Comment (pangyoki) [reply python/paddle/tensor/manipulation.py]: done. The introduction doc of Inplace will be added later.

File: python/paddle/tensor/manipulation.py (hunk 1 above)
Comment (pangyoki) [reply python/paddle/tensor/manipulation.py]: done

File: python/paddle/tensor/manipulation.py (hunk 4 above)
Comment (pangyoki) [reply python/paddle/tensor/manipulation.py]: Have discussed with TPM. The conclusion is that there is no need to add sample code for each inplace API. And the introduction doc of Inplace needs to be added.

**Review by jzhang533 — DISMISSED** [review]
lgtm

File: paddle/fluid/imperative/basic_engine.cc [hunk 8]
```diff
@@ -114,7 +114,9 @@ void BasicEngine::CheckBackwardInputs(const OpBase& op) {
   }
 }
 
-void BasicEngine::PrepareGradAccumulators(const OpBase& op) {
+void BasicEngine::PrepareGradAccumulators(
+    const OpBase& op,
+    const std::vector<std::shared_ptr<GradOpNode>> grad_pending_nodes) {
```
Comment (liym27) [inline paddle/fluid/imperative/basic_engine.cc]: Can `grad_pending_nodes ` passed to this function be as a reference to optimize performance?

File: paddle/fluid/imperative/op_base.h [hunk 9]
```diff
@@ -176,6 +177,8 @@ class OpBase {
   platform::Place place_;
   size_t id_{-1UL};
 
+  std::map<std::string, std::string> inplace_mapping_;
```
Comment (liym27) [inline paddle/fluid/imperative/op_base.h]: What is `inplace_mapping_`?  It seems that `inplace_mapping_` is not used anywhere else.

File: paddle/fluid/imperative/variable_wrapper.h [hunk 10]
```diff
@@ -257,9 +257,10 @@ class VariableWrapper {
 
     auto shared_node = grad_node_.lock();
     if (shared_node != grad_node) {
-      PADDLE_ENFORCE_EQ(
-          shared_node, nullptr,
-          platform::errors::PermissionDenied("Cannot set gradient op twice"));
+      if (shared_node) {
+        VLOG(3) << "The gradient op of Var (" << Name()
+                << ") has been set twice. It's OK if it uses Inplace Strategy.";
+      }
```
Comment (liym27) [inline paddle/fluid/imperative/variable_wrapper.h:268]: If `shared_node` does not use Inplace Strategy, the check `PADDLE_ENFORCE_EQ(shared_node, nullptr)` is necessary.  Better to add the check here.

File: python/paddle/fluid/tests/unittests/test_activation_op.py [hunk 11]
```diff
@@ -261,15 +265,21 @@ def test_errors(self):
...
-            self.assertRaises(TypeError, F.tanh, x_int32)
+            self.assertRaises(TypeError, self.tanh, x_int32)
             # support the input dtype is float16
             x_fp16 = paddle.fluid.data(
                 name='x_fp16', shape=[12, 10], dtype='float16')
-            F.tanh(x_fp16)
+            self.tanh(x_fp16)
+
+
+class TestTanh_InplaceAPI(TestTanhAPI):
```
Comment (liym27) [inline python/paddle/fluid/tests/unittests/test_activation_op.py]: `TestTanh_InplaceAPI `  -> `TestTanhInplaceAPI` to follow the standard CamelCase in Python.

The same below.

**Review by liym27 — COMMENTED** [review]
One suggestion but Not important in this PR. 

I heard from you that there is a performance degradation after using inplace strategy.
Maybe adding a flag to enable/disable the inplace strategy is better.
1. If users don't use inplace Ops or just don't want to enable inplace strategy, they can disable it.
2. it can also reduce the risk of introducing inplace strategy if disable inplace strategy by default now temporarily. (Enable inplace strategy by default after performance verification)

File: paddle/fluid/imperative/basic_engine.cc (hunk 8 above)
Comment (pangyoki) [reply paddle/fluid/imperative/basic_engine.cc]: done, thx.

File: paddle/fluid/imperative/op_base.h (hunk 9 above)
Comment (pangyoki) [reply paddle/fluid/imperative/op_base.h]: have deleted the useless variable.

File: python/paddle/fluid/tests/unittests/test_activation_op.py (hunk 11 above)
Comment (pangyoki) [reply python/paddle/fluid/tests/unittests/test_activation_op.py]: done.

Comment (pangyoki) [reply python/paddle/fluid/tests/unittests/test_activation_op.py]: Modified the name of unittest classes. The underlined (`_`) name usually indicates that it is an inplace operation.

**Review by lanxianghit — DISMISSED** [review]

**Review by zhiqiu — DISMISSED** [review]
LGTM

File: paddle/fluid/imperative/variable_wrapper.h (hunk 10 above)
Comment (pangyoki) [reply paddle/fluid/imperative/variable_wrapper.h:263]: will done in next PR

Comment (pangyoki) [reply paddle/fluid/imperative/variable_wrapper.h:263]: will be done in next PR

File: paddle/fluid/imperative/basic_engine.cc [hunk 12]
```diff
@@ -123,29 +125,94 @@ void BasicEngine::PrepareGradAccumulators(const OpBase& op) {
...
+        }
+      } else {
+        // Because Inplace op overwrites the grad_node of the input grad_var. So
+        // only the information of grad_pending_node can be used to find the
+        // grad_node of grad_var.
+        bool find_grad_node_of_var = false;
+        for (auto& grad_pending_node : grad_pending_nodes) {
+          PADDLE_ENFORCE_NOT_NULL(grad_pending_node,
+                                  platform::errors::NotFound(
+                                      "Grad pending node should not be null"));
```
Comment (chenwhql) [inline paddle/fluid/imperative/basic_engine.cc]: we don't recommend using hint like `should not be...`, we can directly tell users `Grad pending node is not found.` or `... is nullptr.`

File: paddle/fluid/imperative/basic_engine.cc [hunk 13]
```diff
@@ -123,29 +125,94 @@ void BasicEngine::PrepareGradAccumulators(const OpBase& op) {
...
+                    << "(" << var.get()
+                    << ") that has grad node with reference count "
+                    << accumulator->RefCnt();
+            break;
+          }
+        }
+        PADDLE_ENFORCE_EQ(
+            find_grad_node_of_var, true,
+            platform::errors::NotFound(
+                "No grad node corresponding to grad var (%s) was found.",
```
Comment (chenwhql) [inline paddle/fluid/imperative/basic_engine.cc]: var -> Tensor? In 2.0 users only know `Tensor`

File: paddle/fluid/imperative/basic_engine.cc [hunk 14]
```diff
@@ -158,6 +225,9 @@ void BasicEngine::PrepareDeps() {
   PADDLE_ENFORCE_EQ(
       accumulators_.empty(), true,
       platform::errors::AlreadyExists("Accumulators must be initialized here"));
+  PADDLE_ENFORCE_EQ(
+      accumulators_with_grad_node_.empty(), true,
+      platform::errors::AlreadyExists("Accumulators must be initialized here"));
```
Comment (chenwhql) [inline paddle/fluid/imperative/basic_engine.cc]: where is `here`?

**Review by chenwhql — DISMISSED** [review]
Maybe we can polish several error messages in later PR.

File: paddle/fluid/imperative/basic_engine.cc (hunk 12 above)
Comment (pangyoki) [reply paddle/fluid/imperative/basic_engine.cc]: done

File: paddle/fluid/imperative/basic_engine.cc (hunk 13 above)
Comment (pangyoki) [reply paddle/fluid/imperative/basic_engine.cc]: done

File: paddle/fluid/imperative/basic_engine.cc (hunk 14 above)
Comment (pangyoki) [reply paddle/fluid/imperative/basic_engine.cc]: have deleted `here`

File: paddle/fluid/imperative/variable_wrapper.h (hunk 10 above)
Comment (pangyoki) [reply paddle/fluid/imperative/variable_wrapper.h:268]: done, thx.

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by phlrain — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by chenwhql — APPROVED** [review]

File: paddle/fluid/imperative/variable_wrapper.h [hunk 15]
```diff
@@ -258,8 +259,13 @@ class VariableWrapper {
     auto shared_node = grad_node_.lock();
     if (shared_node != grad_node) {
       PADDLE_ENFORCE_EQ(
-          shared_node, nullptr,
-          platform::errors::PermissionDenied("Cannot set gradient op twice"));
+          !shared_node || !grad_node->InplaceGradNameMap().empty(), true,
+          platform::errors::PermissionDenied(
+              "Cannot set gradient op twice unless using Inplace Strategy."));
```
Comment (zhiqiu) [inline paddle/fluid/imperative/variable_wrapper.h:264]: Better not write combined condition in PADDLE_ENFORCE_EQ , You can refine it in next PR.

Comment (pangyoki) [reply paddle/fluid/imperative/variable_wrapper.h:264]: will refine it in next PR

**Review by lanxianghit — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Pull request merged**